tests/unit/src/main/scala/tests/debug/MetalsBaseDebugSuite.scala

```diff
@@ -0,0 +1,92 @@
+package tests.debug
+
+import java.net.URI
+import java.nio.file.Files
+import java.nio.file.Paths
+import java.util.concurrent.ExecutorService
+import java.util.concurrent.Executors
+import java.util.concurrent.TimeUnit
+import org.eclipse.lsp4j.InitializeParams
+import org.eclipse.lsp4j.InitializedParams
+import org.eclipse.lsp4j.debug.InitializeRequestArguments
+import tests.BaseSuite
+import tests.FileLayout
+import tests.QuickBuild
+import scala.concurrent.ExecutionContext.fromExecutorService
+import scala.concurrent.ExecutionContextExecutorService
+import scala.concurrent.Future
+import scala.meta.internal.io.PathIO
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.internal.metals.MetalsLanguageServer
+import scala.meta.internal.metals.NoopLanguageClient
+
+protected[debug] abstract class MetalsBaseDebugSuite extends BaseSuite {
+  private val executor: ExecutorService = Executors.newCachedThreadPool()
+  protected implicit val executionContext: ExecutionContextExecutorService =
+    fromExecutorService(executor)
+  private var metalsServer: MetalsLanguageServer = _
+
+  protected var workspace: URI = _
+  protected var server: DebugProtocolServer = _
+  protected var client: DebugProtocolClient = _
+
+  override def utestBeforeEach(path: Seq[String]): Unit = {
+    if (path.isEmpty) return
+    workspace = PathIO.workingDirectory
+      .resolve("target")
+      .resolve("e2e")
+      .resolve("debug-adapter")
+      .resolve(path.last.replace(' ', '-'))
+      .toURI
+    Files.createDirectories(Paths.get(workspace))
+
+    metalsServer = new MetalsLanguageServer(executionContext)
+    metalsServer.connectToLanguageClient(NoopLanguageClient)
+
+    client = new DebugProtocolClient
+    server = DebugProtocolServer(
+      workspace,
+      metalsServer.compilationScheduler,
+      metalsServer.buildTargets
+    )
+    server.setClient(client)
+  }
+
+  protected final def testDebug(name: String)(
```

Comment:
Although it seems appealing to abstract over the for comprehension boilerplate in `testAsync`, I am concerned that the current implementation of `testDebug` won't be able to cover more advanced user workflows like restart/cancel. I think it would be better if debug tests were  regular slow tests except with additional `debugServer` and `debugClient` helpers in scope.

In the completion test suite, the `check` method has a lot of parameters with default values like `filter: String => Boolean = _ => true`. However, the difference between completions `check` and `testDebug` is that
-  completions are a single synchronous request->response method
- run/debug can be multiple steps with bi-directional asynchronous communication between the client and server